src/logging.cpp

```diff
@@ -239,13 +239,86 @@ namespace BCLog {
     }
 }
 
-void BCLog::Logger::LogPrintStr(const std::string& str, const std::string& logging_function, const std::string& source_file, const int source_line)
+static constexpr uint64_t HOURLY_LOG_QUOTA_IN_BYTES_PER_SOURCE_LOCATION{1024 * 1024};
+
+bool BCLog::Logger::RateLimit(std::string& str, const std::string& logging_function, const SourceLocation& source_location)
+{
+    if (!m_rate_limiting) {
+        // Rate limiting is disabled.
+        return false;
+    }
+
+    const std::chrono::seconds now = GetTime<std::chrono::seconds>();
+    QuotaUsage& quota_usage = m_quota_usage_per_source_location[source_location];
+    // Is the quota exceeded before this log call?
+    bool quota_exceeded_before = quota_usage.m_bytes_logged > HOURLY_LOG_QUOTA_IN_BYTES_PER_SOURCE_LOCATION;
+
+    bool dont_skip = false;
```

Comment:
If i understand the code correctly, `dont_skip` is used to make sure the hourly reset messages are printed.
Maybe the code could be simplified by removing `dont_skip` and calling `LogPrintf` instead for the hourly messages?

Reply:
That is correct!
But this does unfortunately not work because we hold a lock here https://github.com/bitcoin/bitcoin/blob/a2916ae5043436d3fd22acb44c5663227fa7eb90/src/logging.cpp#L317

A recursive call to `LogPrintf` also feels a bit dangerous to me.